### ISSUE TITLE
re-register subscriptions after spark connection dropped

### ISSUE THREAD

**Issue by m-mcgowan** [spark/firmware#278]
https://community.spark.io/t/loosing-spark-subscribe-connection-subscription/6518/3

When the spark connection is established, any registered subscriptions should be re-registered.

At present, subscriptions are registered at the point the code requests them with `Spark.subscribe()`. 

Because of that, I imagine there would be a problem with code like this:

```cpp
SYSTEM_MODE(SEMI_AUTOMATIC)

void setup() {
   // not connected - this won't work.
   Spark.subscribe(...);
}
```

The fix is that the firmware register the subscriptions when connection is made to the cloud (first connection or subsequent.)

**Comment by m-mcgowan** [comment]
Just looking through the code to formulate a solution. `SparkProtocol` currently maintains the subscription filter name and the handler function.  To be able to restore all subscriptions, we need to also maintain the other parameters to `Spark.subscribe()` - namely the event source (`ALL`, `MY_DEVICES`, and a specific <device-id>). 

It's probably simplest to put this in `SparkProtocol` along with the existing subscription details. The `SparkProtocol` class can then send subscribed event details after the handshake.

**Comment by naikrovek** [comment]
This bug, if I understand it correctly, is crushing me at a particular production site.  Wi-Fi is prone to short drops there, and when Wi-Fi drops, the sparks basically forget that they are online at all.  They breathe cyan, but they no longer reach out to the network to do anything from user code.  They show up in /build in the cores list as online, and I can flash them and that fixes everything.  I have lost many days of sensor data by not noticing this before.

**Comment by bkobkobko** [comment]
Hi @naikrovek 

This bug is when one core, let's call it Core A, subscribes to another core's (Core B's) published events.  But then Core A goes offline for some reason (router, internet, etc.) and when it comes back online, it is no longer subscribed to Core B's published events.

So this bug is not about publishing but subscribing.  

If your cores are just publishing and they are breathing cyan but not doing network activity, there is a very good chance that your firmware crashed and the core is running the backup firmware, which is Tinker by default.  When this happens next, see if the cloud thinks that they have the Tinker functions exposed instead of your variables and functions.

**Comment by naikrovek** [comment]
I will check that, thank you.  Is it normal for loss of connectivity to cause a user code crash?

**Comment by bkobkobko** [comment]
No, it is not normal for a loss of connectivity to cause a crash.  In fact my router drops my DHCP lease every 24 hours and my cores go to flashing cyan and back to breathing cyan just fine.  

You can test this in an easy to debug way by just unplugging the upstream connection on your router.

**Comment by pvanallen** [comment]
I'm seeing this exact same issue. I.e. a Spark that has subscribed to an event stops responding to that event when it temporarily loses connection to the WiFi. For example, it is breathing cyan, switches to blinking green, then reconnects and goes back to breathing cyan. After it reconnects, it no longer calls the registered function for that event.

The effect of this bug is to make events essentially impractical in real use cases where WiFi will inevitably drop on occasion.

**Comment by m-mcgowan** [comment]
If you are able to build locally, you can use the feature/hal branch, which contains the fix for this issue.

**Comment by jenesaisdiq** [comment]
This! Fix this in prod.

**Comment by faquin** [comment]
Sorry I am lost in some of Github concepts. I see this issue has been closed, but where can I find in which version it was released?

**Comment by kennethlimcp** [comment]
@faquin i believe this is available from 0.4.x releases

**Comment by m-mcgowan** [comment]
Yes, it's in the 0.4.x release. Already available for the photon, coming to the core today.
